util/wait_group_wrapper.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// WaitGroupWrapper is a wrapper for sync.WaitGroup
+type WaitGroupWrapper struct {
+	sync.WaitGroup
+}
+
+// Run run a function in a goroutine, add 1 to WaitGroup
+// and call done when function return. Please not use panic
+// in the cb function.
+func (w *WaitGroupWrapper) Run(cb func()) {
+	w.Add(1)
+	go func() {
+		cb()
+		w.Done()
+	}()
+}
+
+// RunWithRecover wraps goroutine startup call with force recovery, add 1 to WaitGroup
+// and call done when function return.
+// it will dump current goroutine stack into log if catch any recover result.
+//   exec:      execute logic function.
+//   recoverFn: handler will be called after recover and before dump stack, passing `nil` means noop.
+func (w *WaitGroupWrapper) RunWithRecover(exec func(), recoverFn func(r interface{})) {
+	w.Add(1)
+	go func() {
+		defer func() {
+			r := recover()
+			if recoverFn != nil {
+				recoverFn(r)
```

Comment:
Seems the recoverFn would always run even `exec` didn't panic, is it OK? 

See https://play.golang.org/p/s5gxgKCwFBN

Reply:
Fix

Reply:
I have fixed and push the github. but I don't know why the old version of the code is still displayed.